cypress-io/cypress
# fix: Issue 23486

- Closes https://github.com/cypress-io/cypress/issues/23486

### User facing changelog
- When using the Cypress UI, if the dev server is not set up correctly in the cypress config, users should still be able to see list of spec files they can run.

### Additional details

- The error stems from a missing null check for the component.devServer in the config. This could have happened if component testing was partially set up or devServer dependency was removed after the set up.
- Adding optional chaining to check the existence of the framework on the devServer circumvents the "GraphQL error [GraphQL] Cannot read properties of undefined (reading 'framework')". With optional chaining we will still be checking if the devServer is defined in the framework and if it is undefined it will not throw the error.
- The cypress-in-cypress.cy.ts test manually changes the config file removing the devServer from the component and then checks if the spec list still shows up.

### Steps to test

1. Go to cypress/packages/data-context/src/actions/ProjectActions.ts. On line 534 remove the '?' character between devServer and framework such that it reads as devServer.framework
2. Run the cypress-in-cypress.cy.ts test within the app folder of packages.
3. The 'checks that specs load when devServer configuration is not set in open mode for e2e/CT' test should fail as the spec list is not visible in the DOM.
4. Repeat step 1 BUT add back the '?' character so it reads as devServer?.framework
5. Repeat step 2 and the test should pass now.

### How has the user experience changed?

Before (When config file does not have a devServer):
<img width="1507" alt="Screen Shot 2022-08-29 at 12 12 12 PM" src="https://user-images.githubusercontent.com/65267668/187246937-173ac5d5-9031-433f-90c3-1a6c18dd0b09.png">

After bug fix:
<img width="1507" alt="Screen Shot 2022-08-29 at 12 12 28 PM" src="https://user-images.githubusercontent.com/65267668/187246973-f7c66863-6f6d-4be0-af45-259b2b816cdb.png">


### PR Tasks


- [na] Has the original issue (or this PR, if no issue exists) been tagged with a release in ZenHub? (user-facing changes only)
- [na] Has a PR for user-facing changes been opened in [`cypress-documentation`](https://github.com/cypress-io/cypress-documentation)? 
- [na] Have API changes been updated in the [`type definitions`](https://github.com/cypress-io/cypress/blob/develop/cli/types/cypress.d.ts)?
